master/internal/api_runs.go

```diff
@@ -630,3 +630,129 @@ func (a *apiServer) DeleteRuns(ctx context.Context, req *apiv1.DeleteRunsRequest
+	results, err := archiveUnarchiveAction(ctx, false, req.RunIds, req.ProjectId, req.Filter)
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.UnarchiveRunsResponse{Results: results}, nil
+}
+
+func archiveUnarchiveAction(ctx context.Context, archive bool, runIDs []int32,
+	projectID int32, filter *string,
+) ([]*apiv1.RunActionResult, error) {
+	if len(runIDs) > 0 && filter != nil {
+		return nil, fmt.Errorf("if filter is provided run id list must be empty")
+	}
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var runCandidates []runCandidateResult
+	query := db.Bun().NewSelect().
+		ModelTableExpr("runs AS r").
+		Model(&runCandidates).
+		Column("r.id").
+		Column("r.archived").
+		ColumnExpr("r.experiment_id as exp_id").
+		ColumnExpr("false as is_multitrial").
+		ColumnExpr("r.state IN (?) AS is_terminal", bun.In(model.StatesToStrings(model.TerminalStates))).
+		Join("LEFT JOIN experiments e ON r.experiment_id=e.id").
+		Join("JOIN projects p ON r.project_id = p.id").
+		Join("JOIN workspaces w ON p.workspace_id = w.id").
+		Where("r.project_id = ?", projectID)
+
+	if filter == nil {
+		query = query.Where("r.id IN (?)", bun.In(runIDs))
+	} else {
+		query, err = filterRunQuery(query, filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	query, err = experiment.AuthZProvider.Get().
+		FilterExperimentsQuery(ctx, *curUser, nil, query,
+			[]rbacv1.PermissionType{rbacv1.PermissionType_PERMISSION_TYPE_UPDATE_EXPERIMENT_METADATA})
+	if err != nil {
+		return nil, err
+	}
+
+	err = query.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*apiv1.RunActionResult
+	visibleIDs := set.New[int32]()
+	var validIDs []int32
+	for _, check := range runCandidates {
+		visibleIDs.Insert(check.ID)
+		switch {
+		case check.Archived && archive:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "",
+				Id:    check.ID,
+			})
+		case !check.Archived && !archive:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "",
+				Id:    check.ID,
+			})
+		case !check.IsTerminal:
+			results = append(results, &apiv1.RunActionResult{
+				Error: "Run is not in terminal state.",
+				Id:    check.ID,
+			})
+		default:
+			validIDs = append(validIDs, check.ID)
+		}
+	}
+
+	if filter == nil {
+		for _, originalID := range runIDs {
+			if !visibleIDs.Contains(originalID) {
+				results = append(results, &apiv1.RunActionResult{
+					Error: fmt.Sprintf("Run with id '%d' not found in project with id '%d'", originalID, projectID),
+					Id:    originalID,
+				})
+			}
+		}
+	}
+
+	if len(validIDs) > 0 {
+		var acceptedIDs []int32
+		_, err = db.Bun().NewUpdate().
+			ModelTableExpr("runs as r").
+			Set("archived = ?", archive).
+			Where("id IN (?)", bun.In(validIDs)).
+			Returning("id").
+			Model(&acceptedIDs).
+			Exec(ctx)
+		if err != nil {
+			return nil, err
+		}
+		for _, acceptID := range acceptedIDs {
+			results = append(results, &apiv1.RunActionResult{
```

Comment:
what about for stuff in valid IDs but not in accepted? is that technically possible if something is deleted during this flow.

an aside: I think we should be using db transactions for all these endpoints?

Reply:
second aside: make sure the frontend is ok if it doesn't end up getting a result for every row. or update your tests to check this specifically if they don't already?

Reply:
Don't believe it's an issue if a run is already deleted and we don't show the result in the response. In the case where a filter is used then it would actually be the preferred behaviour

Reply:
I don't understand the other cases then? Why have we been so particular about showing a response if it doesn't matter now?